File: app/src/main/java/ohi/andre/consolelauncher/commands/smartlauncher/developer/GitCommand.java
package ohi.andre.consolelauncher.commands.smartlauncher.developer;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.util.Log;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;

import ohi.andre.consolelauncher.commands.CommandAbstraction;
import ohi.andre.consolelauncher.commands.ExecutePack;

/**
 * Git Command - Basic Git operations for T-UI
 * Supports clone, status, commit, push, pull, and log
 */
public class GitCommand implements CommandAbstraction {

    private static final String TAG = "GitCommand";
    private static final String DEFAULT_HELP = "Git version control operations";

    @Override
    public String exec(ExecutePack pack) throws Exception {
        Context context = pack.context;
        
        ArrayList<String> argsList = new ArrayList<>();
        for (Object arg : pack.args) {
            if (arg != null) {
                argsList.add(arg.toString());
            }
        }

        if (argsList.isEmpty() || "--help".equals(argsList.get(0)) || "-h".equals(argsList.get(0))) {
            return getUsage();
        }

        String command = argsList.get(0).toLowerCase();
        ArrayList<String> parameters = new ArrayList<>(argsList.subList(1, argsList.size()));

        switch (command) {
            case "clone":
                return gitClone(context, parameters);
            case "status":
                return gitStatus(context);
            case "add":
                return gitAdd(context, parameters);
            case "commit":
                return gitCommit(context, parameters);
            case "push":
                return gitPush(context);
            case "pull":
                return gitPull(context);
            case "log":
                return gitLog(context);
            case "diff":
                return gitDiff(context);
            case "branch":
                return gitBranch(context);
            case "checkout":
            case "switch":
                return gitCheckout(context, parameters);
            case "remote":
                return gitRemote(context);
            case "init":
                return gitInit(context);
            case "open":
                return gitOpen(context);
            default:
                return "Unknown git command: " + command + "\n" + getUsage();
        }
    }

    @Override
    public int[] argType() {
        return new int[]{CommandAbstraction.COMMAND};
    }

    @Override
    public int priority() {
        return 5;
    }

    @Override
    public int helpRes() {
        return 0; // We'll return the help text directly in the exec method
    }

    @Override
    public String onArgNotFound(ExecutePack pack, int indexNotFound) {
        return "Missing argument at position " + indexNotFound + "\n" + getUsage();
    }

    @Override
    public String onNotArgEnough(ExecutePack pack, int nArgs) {
        return "Not enough arguments (" + nArgs + " required)\n" + getUsage();
    }

    private String getUsage() {
        return "\n╔══════════════════════════════════════════════════════╗\n" +
               "║                     GIT COMMANDS                      ║\n" +
               "╠══════════════════════════════════════════════════════╣\n" +
               "║  git clone <url>              - Clone repository     ║\n" +
               "║  git status                   - Show status          ║\n" +
               "║  git add <file>               - Stage files          ║\n" +
               "║  git add .                    - Stage all            ║\n" +
               "║  git commit -m \"<message>\"    - Commit changes       ║\n" +
               "║  git push                     - Push to remote       ║\n" +
               "║  git pull                     - Pull from remote     ║\n" +
               "║  git log                      - Show commit log      ║\n" +
               "║  git diff                     - Show changes         ║\n" +
               "║  git branch                   - List branches        ║\n" +
               "║  git checkout <branch>        - Switch branch        ║\n" +
               "║  git remote                   - Show remotes         ║\n" +
               "║  git init                     - Initialize repo      ║\n" +
               "║  git open                     - Open repo in browser ║\n" +
               "╚══════════════════════════════════════════════════════╝\n";
    }

    /**
     * Gets current git directory
     */
    private File getGitDir(Context context) {
        File dir = context.getFilesDir();
        while (dir != null) {
            File gitDir = new File(dir, ".git");
            if (gitDir.exists() && gitDir.isDirectory()) {
                return dir;
            }
            dir = dir.getParentFile();
        }
        return null;
    }

    /**
     * Executes a git command and returns output
     */
    private String executeGit(Context context, String... command) {
        try {
            File gitDir = getGitDir(context);
            if (gitDir == null) {
                return "Not in a git repository. Use 'git clone <url>' or 'git init'.";
            }

            Process process = Runtime.getRuntime().exec(
                mergeArray("git", command), 
                null, 
                gitDir
            );

            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            BufferedReader errorReader = new BufferedReader(new InputStreamReader(process.getErrorStream()));

            StringBuilder output = new StringBuilder();
            String line;

            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }

            while ((line = errorReader.readLine()) != null) {
                output.append(line).append("\n");
            }

            int exitCode = process.waitFor();
            if (exitCode != 0) {
                output.insert(0, "Error (exit code " + exitCode + "):\n");
            }

            return output.toString();
        } catch (Exception e) {
            Log.e(TAG, "Git command failed", e);
            return "Git command failed: " + e.getMessage() + "\nIs Git installed?";
        }
    }

    /**
     * Helper method to merge command array
     */
    private String[] mergeArray(String first, String... rest) {
        String[] result = new String[rest.length + 1];
        result[0] = first;
        System.arraycopy(rest, 0, result, 1, rest.length);
        return result;
    }

    /**
     * Clones a repository
     */
    private String gitClone(Context context, ArrayList<String> args) {
        if (args.isEmpty()) {
            return "Usage: git clone <repository-url> [directory]";
        }

        String url = args.get(0);
        String directory = args.size() > 1 ? args.get(1) : extractRepoName(url);

        File targetDir = new File(context.getFilesDir(), directory);

        try {
            StringBuilder builder = new StringBuilder();
            builder.append("\n");
            builder.append("Cloning repository...\n");
            builder.append("URL: ").append(url).append("\n");
            builder.append("Directory: ").append(directory).append("\n");
            builder.append("─".repeat(50)).append("\n");

            Process process = Runtime.getRuntime().exec(
                mergeArray("git", "clone", url, directory),
                null,
                context.getFilesDir()
            );

            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            String line;
            while ((line = reader.readLine()) != null) {
                builder.append(line).append("\n");
            }

            int exitCode = process.waitFor();
            builder.append("\n");
            builder.append("Exit code: ").append(exitCode).append("\n");

            if (exitCode == 0 && targetDir.exists()) {
                builder.append("✓ Repository cloned successfully!");
            }

            return builder.toString();
        } catch (Exception e) {
            Log.e(TAG, "Clone failed", e);
            return "Clone failed: " + e.getMessage();
        }
    }

    /**
     * Shows repository status
     */
    private String gitStatus(Context context) {
        File repoDir = getGitDir(context);
        if (repoDir == null) {
            return "Not in a git repository.\nUse 'git clone <url>' or 'git init'.";
        }

        String output = executeGit(context, "status", "--short");

        StringBuilder result = new StringBuilder();
        result.append("\n");
        result.append("Repository: ").append(repoDir.getName()).append("\n");
        result.append("─".repeat(50)).append("\n");

        if (output.trim().isEmpty()) {
            result.append("  Working tree clean ✓\n");
        } else {
            result.append(output);
        }

        return result.toString();
    }

    /**
     * Stages files
     */
    private String gitAdd(Context context, ArrayList<String> args) {
        if (args.isEmpty()) {
            return "Usage: git add <file> or git add .";
        }

        String target = ".".equals(args.get(0)) ? "--all" : args.get(0);
        String output = executeGit(context, "add", target);

        return "Files staged.\n" + output;
    }

    /**
     * Commits changes
     */
    private String gitCommit(Context context, ArrayList<String> args) {
        int messageIndex = args.indexOf("-m");
        if (messageIndex == -1 || messageIndex + 1 >= args.size()) {
            return "Usage: git commit -m \"<message>\"";
        }

        String message = args.get(messageIndex + 1);
        if (message.startsWith("\"") && message.endsWith("\"")) {
            message = message.substring(1, message.length() - 1);
        }

        String output = executeGit(context, "commit", "-m", message);

        StringBuilder result = new StringBuilder();
        result.append("Commit created.\n");
        result.append("─".repeat(50)).append("\n");
        result.append(output);

        return result.toString();
    }

    /**
     * Pushes to remote
     */
    private String gitPush(Context context) {
        String output = executeGit(context, "push");

        StringBuilder result = new StringBuilder();
        result.append("Push result:\n");
        result.append("─".repeat(50)).append("\n");
        result.append(output);

        return result.toString();
    }

    /**
     * Pulls from remote
     */
    private String gitPull(Context context) {
        String output = executeGit(context, "pull");

        StringBuilder result = new StringBuilder();
        result.append("Pull result:\n");
        result.append("─".repeat(50)).append("\n");
        result.append(output);

        return result.toString();
    }

    /**
     * Shows commit log
     */
    private String gitLog(Context context) {
        String output = executeGit(context, "log", "--oneline", "-10");

        StringBuilder result = new StringBuilder();
        result.append("\n");
        result.append("Recent Commits (last 10):\n");
        result.append("─".repeat(50)).append("\n");
        
        if (output.trim().isEmpty()) {
            result.append("No commits yet.");
        } else {
            result.append(output);
        }

        return result.toString();
    }

    /**
     * Shows changes
     */
    private String gitDiff(Context context) {
        String output = executeGit(context, "diff", "--stat");

        StringBuilder result = new StringBuilder();
        result.append("Changes:\n");
        result.append("─".repeat(50)).append("\n");
        
        if (output.trim().isEmpty()) {
            result.append("No changes to show.");
        } else {
            result.append(output);
        }

        return result.toString();
    }

    /**
     * Lists branches
     */
    private String gitBranch(Context context) {
        String output = executeGit(context, "branch", "-a");

        StringBuilder result = new StringBuilder();
        result.append("\n");
        result.append("Branches:\n");
        result.append("─".repeat(50)).append("\n");
        
        if (output.trim().isEmpty()) {
            result.append("No branches.");
        } else {
            result.append(output);
        }

        return result.toString();
    }

    /**
     * Switches branch
     */
    private String gitCheckout(Context context, ArrayList<String> args) {
        if (args.isEmpty()) {
            return "Usage: git checkout <branch-name>";
        }

        String branch = args.get(0);
        String output = executeGit(context, "checkout", branch);

        StringBuilder result = new StringBuilder();
        result.append("Switched to branch: ").append(branch).append("\n");
        result.append("─".repeat(50)).append("\n");
        result.append(output);

        return result.toString();
    }

    /**
     * Shows remotes
     */
    private String gitRemote(Context context) {
        String output = executeGit(context, "remote", "-v");

        StringBuilder result = new StringBuilder();
        result.append("\n");
        result.append("Remotes:\n");
        result.append("─".repeat(50)).append("\n");
        
        if (output.trim().isEmpty()) {
            result.append("No remotes configured.");
        } else {
            result.append(output);
        }

        return result.toString();
    }

    /**
     * Initializes new repository
     */
    private String gitInit(Context context) {
        String output = executeGit(context, "init");

        StringBuilder result = new StringBuilder();
        result.append("Repository initialized.\n");
        result.append("─".repeat(50)).append("\n");
        result.append(output);

        return result.toString();
    }

    /**
     * Opens repository URL in browser
     */
    private String gitOpen(Context context) {
        File repoDir = getGitDir(context);
        if (repoDir == null) {
            return "Not in a git repository.";
        }

        // Try to get remote URL
        String output = executeGit(context, "remote", "get-url", "origin");
        String remoteUrl = output.trim();

        if (remoteUrl.isEmpty() || remoteUrl.startsWith("fatal")) {
            return "No remote origin configured.";
        }

        // Convert SSH URL to HTTPS if needed
        String httpsUrl = remoteUrl
                .replace("git@", "")
                .replace(":", "/")
                .replace("github.com/", "github.com/");

        try {
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://" + httpsUrl));
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
            return "Opening: " + httpsUrl;
        } catch (Exception e) {
            Log.e(TAG, "Could not open URL", e);
            return "Could not open URL: " + e.getMessage();
        }
    }

    private String extractRepoName(String url) {
        String trimmed = url.endsWith(".git") ? url.substring(0, url.length() - 4) : url;
        int lastSlash = trimmed.lastIndexOf('/');
        return lastSlash >= 0 ? trimmed.substring(lastSlash + 1) : trimmed;
    }
}

File: tui-smart-launcher/tui/smartlauncher/developer/GitCommand.kt
package tui.smartlauncher.developer

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.util.Log
import tui.smartlauncher.core.CommandHandler
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.io.PrintWriter

/**
 * Git Command - Basic Git operations for T-UI
 * Supports clone, status, commit, push, pull, and log
 */
class GitCommand : CommandHandler {

    companion object {
        private const val TAG = "GitCommand"
    }

    private var context: Context? = null

    override fun getName(): String = "git"

    override fun getAliases(): List<String> = listOf("git", "g")

    override fun getDescription(): String = "Git version control operations"

    override fun getUsage(): String = """
        ╔══════════════════════════════════════════════════════╗
        ║                     GIT COMMANDS                      ║
        ╠══════════════════════════════════════════════════════╣
        ║  git clone <url>              - Clone repository     ║
        ║  git status                   - Show status          ║
        ║  git add <file>               - Stage files          ║
        ║  git add .                    - Stage all            ║
        ║  git commit -m "<message>"    - Commit changes       ║
        ║  git push                     - Push to remote       ║
        ║  git pull                     - Pull from remote     ║
        ║  git log                      - Show commit log      ║
        ║  git diff                     - Show changes         ║
        ║  git branch                   - List branches        ║
        ║  git checkout <branch>        - Switch branch        ║
        ║  git remote                   - Show remotes         ║
        ║  git init                     - Initialize repo      ║
        ║  git open                     - Open repo in browser ║
        ╚══════════════════════════════════════════════════════╝
    """.trimIndent()

    override fun execute(context: Context, args: List<String>): String {
        this.context = context

        if (args.isEmpty() || args[0] == "--help" || args[0] == "-h") {
            return getUsage()
        }

        val command = args[0].lowercase()
        val parameters = args.drop(1)

        return when (command) {
            "clone" -> gitClone(parameters)
            "status" -> gitStatus()
            "add" -> gitAdd(parameters)
            "commit" -> gitCommit(parameters)
            "push" -> gitPush()
            "pull" -> gitPull()
            "log" -> gitLog()
            "diff" -> gitDiff()
            "branch" -> gitBranch()
            "checkout", "switch" -> gitCheckout(parameters)
            "remote" -> gitRemote()
            "init" -> gitInit()
            "open" -> gitOpen()
            else -> "Unknown git command: $command\n${getUsage()}"
        }
    }

    /**
     * Gets current git directory
     */
    private fun getGitDir(): File? {
        var dir = context?.filesDir
        while (dir != null) {
            val gitDir = File(dir, ".git")
            if (gitDir.exists() && gitDir.isDirectory) {
                return dir
            }
            dir = dir.parentFile
        }
        return null
    }

    /**
     * Executes a git command and returns output
     */
    private fun executeGit(vararg command: String): String {
        return try {
            val process = Runtime.getRuntime().exec(arrayOf("git") + command, null, getGitDir())
            val reader = BufferedReader(InputStreamReader(process.inputStream))
            val errorReader = BufferedReader(InputStreamReader(process.errorStream))

            val output = StringBuilder()
            var line: String?

            while (reader.readLine().also { line = it } != null) {
                output.appendLine(line)
            }

            while (errorReader.readLine().also { line = it } != null) {
                output.appendLine(line)
            }

            val exitCode = process.waitFor()
            if (exitCode != 0) {
                output.insert(0, "Error (exit code $exitCode):\n")
            }

            output.toString()
        } catch (e: Exception) {
            "Git command failed: ${e.message}\nIs Git installed?"
        }
    }

    /**
     * Clones a repository
     */
    private fun gitClone(args: List<String>): String {
        if (args.isEmpty()) {
            return "Usage: git clone <repository-url> [directory]"
        }

        val url = args[0]
        val directory = args.getOrNull(1) ?: extractRepoName(url)

        val targetDir = File(context?.filesDir, directory)

        return try {
            val builder = StringBuilder()
            builder.appendLine()
            builder.appendLine("Cloning repository...")
            builder.appendLine("URL: $url")
            builder.appendLine("Directory: $directory")
            builder.appendLine("─".repeat(50))

            val process = Runtime.getRuntime().exec(
                arrayOf("git", "clone", url, directory),
                null,
                context?.filesDir
            )

            val reader = BufferedReader(InputStreamReader(process.inputStream))
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                builder.appendLine(line)
            }

            val exitCode = process.waitFor()
            builder.appendLine()
            builder.appendLine("Exit code: $exitCode")

            if (exitCode == 0 && targetDir.exists()) {
                builder.appendLine("✓ Repository cloned successfully!")
            }

            builder.toString()
        } catch (e: Exception) {
            "Clone failed: ${e.message}"
        }
    }

    /**
     * Shows repository status
     */
    private fun gitStatus(): String {
        val repoDir = getGitDir()
        if (repoDir == null) {
            return "Not in a git repository.\nUse 'git clone <url>' or 'git init'."
        }

        val output = executeGit("status", "--short")

        return buildString {
            appendLine()
            appendLine("Repository: ${repoDir.name}")
            appendLine("─".repeat(50))

            if (output.isBlank()) {
                appendLine("  Working tree clean ✓")
            } else {
                appendLine(output)
            }
        }
    }

    /**
     * Stages files
     */
    private fun gitAdd(args: List<String>): String {
        if (args.isEmpty()) {
            return "Usage: git add <file> or git add ."
        }

        val target = if (args[0] == ".") "--all" else args[0]
        val output = executeGit("add", target)

        return "Files staged.\n$output"
    }

    /**
     * Commits changes
     */
    private fun gitCommit(args: List<String>): String {
        val messageIndex = args.indexOf("-m")
        if (messageIndex == -1 || messageIndex + 1 >= args.size) {
            return "Usage: git commit -m \"<message>\""
        }

        val message = args[messageIndex + 1].removeSurrounding("\"")

        val output = executeGit("commit", "-m", message)

        return buildString {
            appendLine("Commit created.")
            appendLine("─".repeat(50))
            appendLine(output)
        }
    }

    /**
     * Pushes to remote
     */
    private fun gitPush(): String {
        val output = executeGit("push")

        return buildString {
            appendLine("Push result:")
            appendLine("─".repeat(50))
            appendLine(output)
        }
    }

    /**
     * Pulls from remote
     */
    private fun gitPull(): String {
        val output = executeGit("pull")

        return buildString {
            appendLine("Pull result:")
            appendLine("─".repeat(50))
            appendLine(output)
        }
    }

    /**
     * Shows commit log
     */
    private fun gitLog(): String {
        val output = executeGit("log", "--oneline", "-10")

        return buildString {
            appendLine()
            appendLine("Recent Commits (last 10):")
            appendLine("─".repeat(50))
            appendLine(output.ifBlank { "No commits yet." })
        }
    }

    /**
     * Shows changes
     */
    private fun gitDiff(): String {
        val output = executeGit("diff", "--stat")

        return buildString {
            appendLine("Changes:")
            appendLine("─".repeat(50))
            appendLine(output.ifBlank { "No changes to show." })
        }
    }

    /**
     * Lists branches
     */
    private fun gitBranch(): String {
        val output = executeGit("branch", "-a")

        return buildString {
            appendLine()
            appendLine("Branches:")
            appendLine("─".repeat(50))
            appendLine(output.ifBlank { "No branches." })
        }
    }

    /**
     * Switches branch
     */
    private fun gitCheckout(args: List<String>): String {
        if (args.isEmpty()) {
            return "Usage: git checkout <branch-name>"
        }

        val branch = args[0]
        val output = executeGit("checkout", branch)

        return buildString {
            appendLine("Switched to branch: $branch")
            appendLine("─".repeat(50))
            appendLine(output)
        }
    }

    /**
     * Shows remotes
     */
    private fun gitRemote(): String {
        val output = executeGit("remote", "-v")

        return buildString {
            appendLine()
            appendLine("Remotes:")
            appendLine("─".repeat(50))
            appendLine(output.ifBlank { "No remotes configured." })
        }
    }

    /**
     * Initializes new repository
     */
    private fun gitInit(): String {
        val output = executeGit("init")

        return buildString {
            appendLine("Repository initialized.")
            appendLine("─".repeat(50))
            appendLine(output)
        }
    }

    /**
     * Opens repository URL in browser
     */
    private fun gitOpen(): String {
        val repoDir = getGitDir()
        if (repoDir == null) {
            return "Not in a git repository."
        }

        // Try to get remote URL
        val output = executeGit("remote", "get-url", "origin")
        val remoteUrl = output.trim()

        if (remoteUrl.isBlank() || remoteUrl.startsWith("fatal")) {
            return "No remote origin configured."
        }

        // Convert SSH URL to HTTPS if needed
        val httpsUrl = remoteUrl
            .replace("git@", "")
            .replace(":", "/")
            .replace("github.com/", "github.com/")

        return try {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse("https://$httpsUrl"))
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context?.startActivity(intent)
            "Opening: $httpsUrl"
        } catch (e: Exception) {
            "Could not open URL: ${e.message}"
        }
    }

    private fun extractRepoName(url: String): String {
        return url
            .removeSuffix(".git")
            .substringAfterLast("/")
    }
}
